bundle/src/main/java/com/adobe/acs/commons/properties/impl/AllPagePropertiesContentVariableProvider.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * #%L
+ * ACS AEM Commons Bundle
+ * %%
+ * Copyright (C) 2020 Adobe
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.adobe.acs.commons.properties.impl;
+
+import com.adobe.acs.commons.properties.ContentVariableProvider;
+import com.adobe.acs.commons.properties.PropertyConfigService;
+import com.adobe.acs.commons.properties.util.PropertyAggregatorUtil;
+import com.day.cq.wcm.api.Page;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+@Component(service = ContentVariableProvider.class)
+public class AllPagePropertiesContentVariableProvider implements ContentVariableProvider {
+
+    public static final String PAGE_PROP_PREFIX = "page_properties";
+    private static final String INHERITED_PAGE_PROP_PREFIX = "inherited_page_properties";
+
+    @Reference
+    private PropertyConfigService propertyConfigService;
+
+    @Override
+    public void addProperties(Map<String, Object> map, Page page) {
+        // Add current page properties
+        PropertyAggregatorUtil.addPagePropertiesToMap(map, page, PAGE_PROP_PREFIX, propertyConfigService);
+
+        // Add inherited page properties
+        Page parent = page.getParent();
+        while (parent != null) {
+            Map<String, Object> inheritedMap = new HashMap<>();
+            PropertyAggregatorUtil.addPagePropertiesToMap(inheritedMap, parent, INHERITED_PAGE_PROP_PREFIX, propertyConfigService);
+            Set<Map.Entry<String, Object>> entries = inheritedMap.entrySet();
+            for (Map.Entry<String, Object> entry : entries) {
+                if (shouldAddInherited(map, entry.getKey())) {
+                    map.put(entry.getKey(), entry.getValue());
+                }
+            }
+            parent = parent.getParent();
+        }
+    }
+
+    /**
+     * Check to see if the current inherited property key is already contained in the set of properties.
+     *
+     * @param map          current map of properties
+     * @param propertyName current property name
+     * @return whether the map contains a local page property or an inherited property
+     */
+    private boolean shouldAddInherited(Map<String, Object> map, String propertyName) {
+        return !map.containsKey(propertyName.replace(INHERITED_PAGE_PROP_PREFIX, PAGE_PROP_PREFIX))
```

Comment:
Does this mean that if a property has a page-level value then we wont set the inherited value?  What does the author do in a situation where a value might be on the page, or might be inherited?

Reply:
This will put together all the local and inherited properties with priority on the local properties (basically the same as `HierarchyNodeInheritanceValueMap`). They will all be prefixed so the author will either be accessing `inheritedPageProperties` or `pageProperties` (explicitly called out). The one difference here between accessing the properties from the FE is that while `${inheritedPageProperties.pageTitle}` does an actual check for a local property before iterating up the tree - the templated dialogs stuff is more explicit - if a local property is set for `pageTitle` then `inheritedPageProperties.pageTitle` won't return anything. This is to make the overall process easier for extension and have 'simple' key/value mappings authors can use directly. 
In the case you provided they'd have to specify whether the property is coming from the current page or inherited - there's no actual logic associated with the replacement other than string matching once the properties are aggregated - i think this makes the process more performant as we don't have to resolve potentially multiple values for the same token.
Also if it's a requirement for the author to have a property that can come from either place there's always the option for their dev team to extend this for their specific use-case :D

Reply:
ok, it just seems "normal" to me for an "inherited" value to treat a local value with priority, but maybe that's just me.  What scenarios are you thinking of where an author wants an inherited value but NOT a page value?  Perhaps that will help me get on board.

Reply:
I agree that it seems normal for that, it's just hard to do without extra processing. The problem currently is that I can't aggregate the full list of tokens for the user to use without differentiating between inherited and local properties - it's why they have different prefixes and there's the check for 'if there isn't a local version of this property'
would it make more sense to have this tool do the following instead? (it would be less performant but hopefully less confusing)
have `AllPagePropertiesContentVariableProvider` provide a list of all page variables (inherited or otherwise) just simply be `page.propertyName` and when it's referenced on a page the replacement tool looks for local properties and then inherited? i could do that or just have `inherited` look for local first. 
overall the problem to solve would be that currently the aggregation is done before the reference is resolved.

Reply:
not completely following, but my perspective is:
- I think it's fine that `page_properties` and `inherited_page_properties` have separate keys
- I dont see why it's a problem for both to have the same value if the value they are looking for lives directly on the current page.  As a user, that's what I would expect.

Is the solution simply:
- Remove `shouldAddInherited`
- Update `inherited_page_properties` to include properties that are on the current page as the highest priority